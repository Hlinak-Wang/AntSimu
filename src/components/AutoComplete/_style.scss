.input-autocomplete {
  position: relative;
}

.suggestion-list {
  position: absolute;
  width: 100%;
  box-shadow: 0 0 6px 1px $gray-7;
  font-weight: 300; 
  background: $white;
  list-style: none;
  padding: 0;
  z-index: 2;
}

.suggestion-item {
  padding: 5px 10px;
  margin: 0;
  cursor: pointer;
}

.hightLighted {
  background: $gray-3;
  color: $black
}

.same-suggestion {
  background: lighten($primary, 30%);
  font-weight: 600;
  color: $black
}