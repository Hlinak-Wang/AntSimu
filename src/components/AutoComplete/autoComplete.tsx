import React, { FC, KeyboardEvent, useState, ReactElement, useRef, ChangeEvent } from 'react';
import Animate from '../Animate/animate';
import classnames from 'classnames';
import Input, { InputProps } from '../Input/input';
import useClickOutSide from '../../hooks/useClickOutSide';
import debounce from '../../hooks/debounce'
export type OptionType<T = {}> = T & {value: string;}
export type FetchHandle = (keyword: string) => OptionType[] | Promise<OptionType[]>

export interface IAutoComplete extends Omit<InputProps, 'onSelect'> {
  fetchSuggestion: FetchHandle
  onSelect?: (item: OptionType) => void
  renderOption?: (item: OptionType) => ReactElement
}

export const useFetch = (fetchSuggestion: FetchHandle, delay: number, imm:boolean) => {
  const [isLoad, setLoading] = useState(false)
  const [fetchActive, setActive] = useState(false)
  const [suggestions, setSuggestion] = useState<OptionType[]>([])
  
  const goFetch =(value: string) => {
    // 当Input.value有值时
    if (value) {
      const results = fetchSuggestion(value);
      if (results instanceof Promise) {
        setActive(true)
        setLoading(true);
        results.then(data => {
          if (data.length !== 0) {
            setLoading(false)
            setSuggestion([...data]);
          }
        }).catch(err => {
          console.log(err)
        })
      } else {
        setLoading(false);
        setActive(true)
        setSuggestion([...results]);
      }
    }
  }

  const fetching = debounce(goFetch, delay, imm)
  return { suggestions, isLoad, fetching, fetchActive, setActive }
}

export const AutoComplete:FC<IAutoComplete> = (props) => {
  const {
    fetchSuggestion,
    onSelect,
    onChange,
    value,
    renderOption,
    ...restProps
  } = props;
  
  const [inputValue, setInputValue] = useState(value as string);
  const [hightLight, setHightLight] = useState(-1);
  const [inputActive, setInputActive] = useState(false);
  const componentRef = useRef<HTMLDivElement | null>(null);
  const inputRef = useRef<HTMLInputElement | null>(null);
  const { suggestions, isLoad, fetching, fetchActive, setActive } = useFetch(fetchSuggestion, 200, true)

  useClickOutSide(componentRef, () => {setActive(false)})

  const handleSelect = (item: OptionType) => {
    console.log("asdfzxc")
    setInputValue(item.value);
    onSelect && onSelect(item)
  }

  const handleHightLight = (newHightLight: number) => {
    if (newHightLight > suggestions.length - 1) {
      setHightLight(suggestions.length - 1);
    } else if (newHightLight < 0) {
      setHightLight(0);
    } else {
      setHightLight(newHightLight);
    }
  }

  const handleKeyboard = (e: KeyboardEvent<HTMLInputElement>) => {
    switch (e.keyCode) {
      case 13:  // enter
        setInputValue(suggestions[hightLight].value);
        setHightLight(-1);
        break;
      case 38:  // up
        handleHightLight(hightLight - 1);
        break;
      case 40:  //down
        handleHightLight(hightLight + 1);
        break;
      case 27:  //esc
        setHightLight(-1);
        break;
      default:
        break;
    }
  }

  const generateDropDown = () => {
    const offsetTop = componentRef.current && componentRef.current.clientHeight + 6;
    return (
      // <Animate
      //   in={inputActive && fetchActive}
      //   timeout={300}
      //   classNames={`test-top`}
      //   unmountOnExit
      // >
        <ul className="suggestion-list" style={{top: `${offsetTop}px`}}>
          {isLoad 
          ? <li><h1>loading</h1></li>
          : suggestions.map((suggestion, index) => {
            const cls = classnames('suggestion-item', {
              'hightLighted': index === hightLight,
              'same-suggestion': suggestion.value === inputValue
            })
            return (
              <li 
                key={index} 
                className={cls} 
                onMouseEnter={() => {console.log("asdf"); setHightLight(index)}} 
                onClick={() => {handleSelect(suggestion)}}
              >
                {renderOption ? renderOption(suggestion) : suggestion.value}
              </li>
            )
          })
          }
        </ul>
      //</Animate>
    )
  }

  const handleFocus = () => {
    setInputActive(true);
  }

  const handleBlur = () => {
    setInputActive(false);
  }

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    
    setInputValue(e.target.value)
    fetching(e.target.value)
    onChange && onChange(e)
    console.log(fetchActive)
  }

  return (
    <div className="input-autocomplete" ref={componentRef}>
      <Input 
        ref={inputRef} 
        onFocus={handleFocus}
        onBlur={handleBlur}
        value={inputValue} 
        onChange={handleChange} 
        onKeyDown={handleKeyboard} 
        {...restProps} 
      />
      {inputActive && fetchActive && generateDropDown()} 
    </div>
  )
}

export default AutoComplete;