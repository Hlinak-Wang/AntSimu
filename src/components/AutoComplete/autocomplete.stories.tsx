import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import AutoComplete, { OptionType } from './autoComplete';

const localAuto = () => {
  type DataType = {value: string; id: number};
  
  const a: DataType[] = [
    {value: 'apple', id: 1}, 
    {value: 'book', id: 2}, 
    {value: 'cat', id: 3}, 
    {value: 'dog', id: 4}, 
    {value: 'egg', id: 5},
  ];

  const handleFetch = (keyword: string) => {
    return a.filter(v => v.value.includes(keyword))
  } 

  return <AutoComplete fetchSuggestion={handleFetch} onSelect={action('selected')} value="a"/>

}

const netAuto = () => {
  type DataType = {value: string};

  const handleFetch = (keyword: string) => {
    // send fetch request
    return fetch(`https://api.github.com/search/users?q=${keyword}`)
    .then(res => res.json())
    .then(({ items }) => {
      return items.slice(0, 10).map((item: { login: any; }) => ({value: item.login, ...item}))
    })
  }

  const renderOption = (item: OptionType) => {
    const options = item as OptionType<DataType>;
    return (
      <>
      <span>name: {options.value}</span>
      </>
    )
  }
  return <AutoComplete fetchSuggestion={handleFetch} onSelect={action('selected')} renderOption={renderOption}/>
  
}

storiesOf('autocomplete', module)
  .add('同步请求', localAuto)
  .add('异步请求', netAuto)