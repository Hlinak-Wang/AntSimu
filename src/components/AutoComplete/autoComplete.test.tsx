import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import AutoComplete from './autoComplete'

const {shallow, mount}=Enzyme
Enzyme.configure({ adapter: new Adapter() })

type DataType = {value: string; id: number};
const a: DataType[] = [
  {value: 'apple', id: 1}, 
  {value: 'book', id: 2}, 
  {value: 'cat', id: 3}, 
  {value: 'dog', id: 4}, 
  {value: 'egg', id: 5},
];
describe(('autocomplete test'), () => {
  let event:any;
  beforeEach(() => {
    event = {
      onSelect: jest.fn(),
      onChange: jest.fn()
    }
  })
  it('should act as input as normal', () => {
    const handleFetch = (keyword: string) => {
      return a.filter(v => v.value.includes(keyword))
    }
    const wrapp = mount(<AutoComplete fetchSuggestion={handleFetch} {...event}/>)
    const input = wrapp.find('input')
    // ensure empty initial state
    expect(input.prop('value')).toEqual('')
    expect(wrapp).toMatchSnapshot()
  })
  it('should show settled fetch information', () => {
    const handleFetch = (keyword: string) => {
      return a.filter(v => v.value.includes(keyword))
    }
    const wrapp = mount(<AutoComplete fetchSuggestion={handleFetch} {...event} />)
    const input = wrapp.find('input')
    
    input.simulate('focus')
    input.simulate('change', {target: {value: "a"}})
    expect(event.onChange).toHaveBeenCalled()

    const list = wrapp.find(".suggestion-list")
    expect(list.find("li").length).toEqual(2)

    list.childAt(0).simulate('click')
    expect(event.onSelect).toHaveBeenCalledWith({value: 'apple', id: 1})
    console.log(input.debug())
  })
  it('test fetch online', async () => {
    const handleFetch = (keyword: string) => {
      // send fetch request
      return fetch(`https://api.github.com/search/users?q=${keyword}`)
      .then(res => res.json())
      .then(({ items }) => {
        return items.slice(0, 10).map((item: { login: any; }) => ({value: item.login, ...item}))
      })
    }

    const wrapp = mount(<AutoComplete fetchSuggestion={handleFetch} onSelect={event.onSelect}/> )
    const input = wrapp.find('input')
    expect(wrapp).toMatchSnapshot()
    
    input.simulate('focus')
    input.simulate('change', {target: {value: "a"}})
  })
})