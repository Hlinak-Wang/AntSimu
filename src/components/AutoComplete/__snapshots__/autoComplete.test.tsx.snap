// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete test should act as input as normal 1`] = `ReactWrapper {}`;

exports[`autocomplete test test fetch online 1`] = `ReactWrapper {}`;
